benchmarks/persistence/persistence.py

```diff
@@ -0,0 +1,205 @@
+
+    execution_date_time: datetime = field(validator=instance_of(datetime))
+    """The date and time when the benchmark was executed."""
+
+    branch: str = field(validator=instance_of(str), init=False)
+    """The branch currently checked out."""
+
+    @branch.default
+    def _default_branch(self) -> str:
+        repo = Repo(search_parent_directories=True)
+        current_branch = repo.active_branch.name
+        return current_branch
+
+    def _string_sanitizer(self, string: str) -> str:
+        """Replace not allowed chars for filesystems in a string."""
+        ALLOWED_CHARS = (
+            "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-."
+        )
+        return "".join([char if char in ALLOWED_CHARS else "-" for char in string])
+
+    def get_path(self, strategy: PathStrategy) -> str:
+        """Construct the path of a result object.
+
+        Create the path depending on the chosen strategy.
+        It will make use of the experiment identifier,
+        the branch, the BayBE-version, the start date without time and
+        the commit hash.
+
+        Args:
+            strategy: The strategy to construct the path.
+
+        Returns:
+            The path to persist the object. Can be a file name or
+            a folder path as a string.
+        """
+        separator = "/" if strategy == PathStrategy.Hierarchical else "_"
+
+        file_usable_date = self.execution_date_time.strftime("%Y-%m-%d")
+        components = [
+            self.benchmark_name,
+            self.branch,
+            self.latest_baybe_tag,
+            file_usable_date,
+            self.commit_hash,
+        ]
+
+        sanitized_components = [
+            self._string_sanitizer(component) for component in components
+        ]
+        path = separator.join(sanitized_components) + separator + "result.json"
+
+        return path
+
+
+class PersistenceHandlingInterface(Protocol):
+    """Interface for interacting with storage."""
+
+    def write_object(self, path: PathConstructor, object: dict) -> None:
+        """Store a JSON serializable dict according to the path.
+
+        Args:
+            path: The path of the object to be persisted
+            object: The object to be persisted.
+        """
+
+
+@define
+class S3PersistenceHandler:
+    """Class for persisting objects in an S3 bucket."""
+
+    _bucket_name: str = field(validator=instance_of(str), init=False)
+    """The name of the S3 bucket where the results are stored."""
+
+    _object_session = boto3.session.Session()
+    """The boto3 session object. This will load the respective credentials
+    from the environment variables within the container."""
+
+    @_bucket_name.default
+    def _default_bucket_name(self) -> str:
+        """Get the bucket name from the environment variables."""
+        if "BAYBE_PERFORMANCE_PERSISTENCE_PATH" not in os.environ:
+            raise ValueError("No S3 bucket name provided.")
```

Comment:
Add the information that this needs to be set via Environment Variable to this error message.